components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt

```diff
@@ -0,0 +1,109 @@
+import java.io.File
+
+/**
+ * A [CrashReporterService] implementation for recording metrics with Glean.  The purpose of this
+ * crash reporter is to collect crash count metrics by capturing [Crash.UncaughtExceptionCrash] and
+ * [Crash.NativeCodeCrash] events and record to the respective
+ * [mozilla.components.service.glean.private.CounterMetricType].
+ */
+class GleanCrashReporterService(val context: Context) : CrashReporterService {
+    companion object {
+        // This file is stored in the application's data directory, so it should be located in the
+        // same location as the application.
+        // The format of this file is simple and uses the keys named below, one per line, to record
+        // crashes.  That format allows for multiple crashes to be appended to the file if, for some
+        // reason, the application cannot run and record them.
+        const val CRASH_FILE_NAME = "glean_crash_counts"
+
+        // These keys correspond to the labels found for crashCount metric in metrics.yaml as well
+        // as the persisted crashes in the crash count file (see above comment)
+        const val UNCAUGHT_EXCEPTION_KEY = "uncaught_exception"
+        const val NATIVE_CODE_CRASH_KEY = "native_code_crash"
+    }
+
+    private val logger = Logger("glean/GleanCrashReporterService")
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal val file = File(context.applicationInfo.dataDir, CRASH_FILE_NAME)
+
+    init {
+        // We only want to record things on the main process because that is the only one in which
+        // Glean is properly initialized.  Checking to see if we are on the main process here will
+        // prevent the situation that arises because the integrating app's Application will be
+        // re-created when prompting to report the crash, and Glean is not initialized there since
+        // it's not technically the main process.
+        if (context.isMainProcess() && file.exists() && file.canRead()) {
+            // Because lib-crash runs in a separate process when report() is called, we need to do
+            // actual recording here because the GleanCrashReporterService constructor is called and
+            // runs in the process where Glean is actually initialized.
+            //
+            // NOTE:  It is important to make sure to initialize Glean before instantiating and
+            // registering the CrashReporter!
+            val lines = file.readLines()
+
+            // It's possible that there was more than one crash recorded in the file so process each
+            // line.
+            lines.forEach { line ->
+                when (line) {
+                    UNCAUGHT_EXCEPTION_KEY -> CrashMetrics.crashCount[UNCAUGHT_EXCEPTION_KEY].add()
+                    NATIVE_CODE_CRASH_KEY -> CrashMetrics.crashCount[NATIVE_CODE_CRASH_KEY].add()
+                }
+            }
+
+            // Clear persisted counts by deleting the file
+            file.delete()
```

Comment:
Can this be hit while we're writing something?

Reply:
no, if we crash in this process, it would be suspended/terminated when the lib-crash process was writing.  I suppose it could be possible to crash while writing if something very weird happened, but that would likely result in a startup-crash loop.